### PR TITLE
Move pyutilib.misc.gc_manager to pyomo.common.gc_manager

### PR DESCRIPTION
**Pull request by mrmundt**
## Summary/Motivation:
In preparation for Pyomo 6.0, we wish to separate PyUtilib. This PR moves the garbage collector manager from PyUtilib to Pyomo.

**TO DO**:

- [x] Create `gc_manager` test(s)

## Changes proposed in this PR:
- Move `pyutilib.misc.gc_manager` to `pyomo.common.gc_manager`

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

**Comment by mrmundt** [comment]
@jsiirola - I'm not sure how you would test this other than just turning it on and off. Is there a way to confirm that it's doing what is expected?

**Comment by jsiirola** [comment]
I think I would just do a simple test for nested use:
```python
import gc

class TestPauseGC(unittest.TestCase):
    def test_gc_disable(self):
        self.assertTrue(gc.isenabled())
        pgc = PauseGC()
        self.assertFalse(gc.isenabled())
        with PauseGC():
            self.assertFalse(gc.isenabled())
        self.assertFalse(gc.isenabled())
        pgc.close()
        self.assertTrue(gc.isenabled())

        self.assertTrue(gc.isenabled())
        with PauseGC():
            self.assertFalse(gc.isenabled())
            pgc = PauseGC()
            self.assertFalse(gc.isenabled())
            pgc.close()
            self.assertFalse(gc.isenabled())
        self.assertTrue(gc.isenabled())
```

File: pyomo/common/gc_manager.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    def __init__(self, pause=True):
+        self.reenable_gc = False
+        if pause:
+            try:
+                self.reenable_gc = gc.isenabled()
+                gc.disable()
+            except NotImplementedError:
+                # This will only happen if the Python implementation
+                # doesn't support disabling the GC.
+                pass
```
Comment (jsiirola) [inline pyomo/common/gc_manager.py:46]: Let' delete this try/except.  It is only needed for Jython, which we don't support (and given that Jython only supports Python 2.7, our upcoming end of 2.7 support will mean that we will not add Jython support)

**Comment by mrmundt** [comment]
@jsiirola - Changes added.

**Review by jsiirola — APPROVED** [review]

**Pull request merged**